core/src/main/java/feast/core/job/JobUpdateTask.java

```diff
@@ -73,147 +74,131 @@ public JobUpdateTask(
-              "Job status updated: changed from %s to %s",
-              job.getStatus(),
-              newJobStatus);
-        }
-        job.setStatus(newJobStatus);
-        return job;
+
+    if (currentJob.isEmpty()) {
+      submittedJob = executorService.submit(this::createJob);
+    } else {
+      Job job = currentJob.get();
+
+      if (featureSetsChangedFor(job)) {
+        submittedJob = executorService.submit(() -> updateJob(job));
       } else {
-        submittedJob =
-            executorService.submit(() -> updateJob(currentJob.get(), featureSets, store));
+        return updateStatus(job);
       }
-    } else {
-      String jobId = createJobId(source.getId(), store.getName());
-      submittedJob = executorService.submit(() -> startJob(jobId, featureSets, sourceSpec, store));
     }
 
-    Job job = null;
     try {
-      job = submittedJob.get(getJobUpdateTimeoutSeconds(), TimeUnit.SECONDS);
+      return submittedJob.get(getJobUpdateTimeoutSeconds(), TimeUnit.SECONDS);
     } catch (InterruptedException | ExecutionException | TimeoutException e) {
+      Source source = Source.fromProto(sourceSpec);
       log.warn("Unable to start job for source {} and sink {}: {}", source, store, e.getMessage());
+      return null;
+    } finally {
       executorService.shutdownNow();
     }
-    return job;
+  }
+
+  boolean featureSetsChangedFor(Job job) {
+    Set<String> existingFeatureSetsPopulatedByJob =
+        job.getFeatureSets().stream().map(FeatureSet::getId).collect(Collectors.toSet());
+    Set<String> newFeatureSetsPopulatedByJob =
+        featureSets.stream()
+            .map(fs -> FeatureSet.fromProto(fs).getId())
+            .collect(Collectors.toSet());
+
+    return !newFeatureSetsPopulatedByJob.equals(existingFeatureSetsPopulatedByJob);
+  }
+
+  private Job createJob() {
+    Source source = Source.fromProto(sourceSpec);
+    String jobId = createJobId(source.getId(), store.getName());
+    return startJob(jobId);
   }
 
   /** Start or update the job to ingest data to the sink. */
-  private Job startJob(
-      String jobId,
-      List<FeatureSetProto.FeatureSet> featureSetProtos,
-      SourceProto.Source source,
-      StoreProto.Store sinkSpec) {
-
-    List<FeatureSet> featureSets =
-        featureSetProtos.stream()
-            .map(
-                fsp ->
-                    FeatureSet.fromProto(
-                        FeatureSetProto.FeatureSet.newBuilder()
-                            .setSpec(fsp.getSpec())
-                            .setMeta(fsp.getMeta())
-                            .build()))
-            .collect(Collectors.toList());
+  private Job startJob(String jobId) {
+
     Job job =
         new Job(
             jobId,
             "",
             jobManager.getRunnerType().name(),
-            Source.fromProto(source),
-            Store.fromProto(sinkSpec),
-            featureSets,
+            Source.fromProto(sourceSpec),
+            Store.fromProto(store),
```

Comment:
Nitpick here but we should probably make our naming consistent:
- suffix `Spec` (ie `sourceSpec` & `storeSpec`) for protobuf versions of the object.
- reserve non suffixed `source` & `store` for native version of the object

Reply:
Agreed. That was the original intention, but it seems like some kind of drift has occurred.

Reply:
I can do that. In light of @zhilingc's [answer above][1] I'll update the PR with removing the Executor usage in `JobUpdateTask`, so I'll add a commit making the naming consistent too.

I wasn't fully noticing this intended convention myself—"feature set spec" (especially since v0.4) I got because of `spec` and `meta` as children of `FeatureSet` in the schema, and generally the distinction between the "registry specification" and a concrete set of values, an "instance" of a feature set you might say. Makes sense that the same logic applies to sources and stores, and hopefully as more domain model classes are getting defined now we will start to see fewer of the "specs" in core logic and that naming will make them stand out.

[1]: https://github.com/gojek/feast/pull/650#discussion_r415472243